README.md
# Ex_RedisGraph

A [RedisGraph](https://redis.io/docs/stack/graph/) client library in Elixir with support for Cypher query building.

The library is developed on top of an existing [library](https://github.com/crflynn/redisgraph-ex),
written by Christopher Flynn, and provides additional functionality with refactored codebase to support
[RedisGraph result set](https://redis.io/docs/stack/graph/design/client_spec/).

Library is publishd to [Hex](https://hex.pm/packages/ex_redisgraph).

Documentation for the library can be found [here](https://hexdocs.pm/ex_redisgraph/RedisGraph.html).

**Note:** if you are new to Elixir, please follow the **Prerequisites.txt** file located in the root of the project.

## Installation
Add the `:ex_redisgraph` dependency to your `mix.exs` file. 
```elixir
defp deps() do
  [
    {:ex_redisgraph, "~> 0.1.0"}
  ]
end
```

To run `RedisGraph` locally with Docker, use

```bash
docker run -p 6379:6379 -it --rm redis/redis-stack-server
```

## Examples
Here is a simple example of how to use the library:

```elixir
alias RedisGraph.{Query, Graph, QueryResult}

# Create a connection using Redix
{:ok, conn} = Redix.start_link("redis://localhost:6379")

# Create a graph
graph = Graph.new(%{
  name: "social"
})

{:ok, query} = Query.new()
                |> Query.create()
                |> Query.node(:n, ["Person"], %{age: 30, name: "John Doe", works: true})
                |> Query.relationship_from_to(:r, "TRAVELS_TO", %{purpose: "pleasure"})
                |> Query.node(:m, ["Place"], %{name: "Japan"})
                |> Query.return(:n)
                |> Query.return_property(:n, "age", :Age)
                |> Query.return(:m)
                |> Query.build_query()

# query will hold
# "CREATE (n:Person {age: 30, name: 'John Doe', works: true})-[r:TRAVELS_TO {purpose: 'pleasure'}]->(m:Place {name: 'Japan'}) RETURN n, n.age AS Age, m

# Execute the query
{:ok, query_result} = RedisGraph.query(conn, graph.name, query)

# Get result set
result_set = Map.get(query_result, :result_set)
# result_set will hold
# [
#   [
#     %RedisGraph.Node{
#       id: 2,
#       alias: :n,
#       labels: ["Person"],
#       properties: %{age: 30, name: "John Doe", works: true}
#     },
#     30,
#     %RedisGraph.Node{
#       id: 3,
#       alias: :m,
#       labels: ["Place"],
#       properties: %{name: "Japan"}
#     }
#   ]
# ]
```
## License

Ex_RedisGraph is licensed under [MIT](https://github.com/AlexSandro19/redisgraph-ex-lib/blob/master/licenses/LICENSE_AlexSandro19.txt).

CHANGELOG.md
# Changelog

## v0.1.0

* Initial release of the library
* Support for Cypher query building
* Communication with RedisGraph
* Representation of real data through Graph/Node/Relationhip structures.


test/redis_graph_test.exs
defmodule RedisGraphTest do
  alias RedisGraph.Relationship
  alias RedisGraph.Graph
  alias RedisGraph.Node
  alias RedisGraph.QueryResult

  use ExUnit.Case

  @redis_address "redis://localhost:6379"

  # def build_sample_graph() do
  #   graph =
  #     Graph.new(%{
  #       name: "social"
  #     })

  #   john =
  #     Node.new(%{
  #       label: "person",
  #       properties: %{
  #         name: "John Doe",
  #         age: 33,
  #         gender: "male",
  #         status: "single"
  #       }
  #     })

  #   {graph, john} = Graph.add_node(graph, john)

  #   japan =
  #     Node.new(%{
  #       label: "country",
  #       properties: %{
  #         name: "Japan"
  #       }
  #     })

  #   {graph, japan} = Graph.add_node(graph, japan)

  #   relationship =
  #     Relationship.new(%{
  #       src_node: john,
  #       dest_node: japan,
  #       relation: "visited"
  #     })

  #   {:ok, graph} = Graph.add_relationship(graph, relationship)
  #   graph
  # end

  # test "creates an execution plan" do
  #   {:ok, conn} = Redix.start_link(@redis_address)

  #   sample_graph = build_sample_graph()

  #   {:ok, commit_result} = RedisGraph.commit(conn, sample_graph)
  #   %QueryResult{} = commit_result

  #   assert Map.get(commit_result.statistics, "Nodes created") == "2"
  #   assert Map.get(commit_result.statistics, "Relationships created") == "1"

  #   q = "MATCH (p:person)-[]->(j:place {purpose:\"pleasure\"}) RETURN p"
  #   {:ok, plan} = RedisGraph.execution_plan(conn, sample_graph.name, q)

  #   assert plan ==
  #            [
  #              "Results",
  #              "    Project",
  #              "        Conditional Traverse | (j:place)->(p:person)",
  #              "            Filter",
  #              "                Node By Label Scan | (j:place)"
  #            ]

  #   # cleanup
  #   {:ok, delete_result} = RedisGraph.delete(conn, sample_graph.name)
  #   assert is_binary(Map.get(delete_result.statistics, "Query internal execution time"))
  # end

  # test "gets results from call procedures" do
  #   {:ok, conn} = Redix.start_link("redis://localhost:6379")

  #   sample_graph = build_sample_graph()

  #   {:ok, commit_result} = RedisGraph.commit(conn, sample_graph)
  #   %QueryResult{} = commit_result

  #   assert Map.get(commit_result.statistics, "Nodes created") == "2"
  #   assert Map.get(commit_result.statistics, "Relationships created") == "1"

  #   {:ok, labels_result} = RedisGraph.labels(conn, sample_graph.name)

  #   labels =
  #     labels_result
  #     |> Enum.at(1)
  #     |> Enum.map(fn element -> element |> Enum.at(0) |> Enum.at(1) end)

  #   assert "person" in labels
  #   assert "country" in labels

  #   {:ok, property_keys_result} = RedisGraph.property_keys(conn, sample_graph.name)

  #   property_keys =
  #     property_keys_result
  #     |> Enum.at(1)
  #     |> Enum.map(fn element -> element |> Enum.at(0) |> Enum.at(1) end)

  #   assert "age" in property_keys
  #   assert "gender" in property_keys
  #   assert "name" in property_keys
  #   assert "status" in property_keys

  #   {:ok, relationship_types_result} = RedisGraph.relationship_types(conn, sample_graph.name)

  #   relationship_types =
  #     relationship_types_result
  #     |> Enum.at(1)
  #     |> Enum.map(fn element -> element |> Enum.at(0) |> Enum.at(1) end)

  #   assert "visited" in relationship_types

  #   {:ok, delete_result} = RedisGraph.delete(conn, sample_graph.name)
  #   assert is_binary(Map.get(delete_result.statistics, "Query internal execution time"))
  # end

  setup_all do
    {atom, process_id} = Redix.start_link(@redis_address)
    graph = Graph.new(%{name: "test"})
    {atom, %{conn: process_id, graph: graph}}
  end

  setup context do
    on_exit(fn ->
      conn = context[:conn]
      graph = context[:graph]
      RedisGraph.delete(conn, graph.name)
    end)

    :ok
  end

  describe "RedisGraph.command" do
    test "test that QueryResult is being returned with correct data",
         %{conn: conn, graph: graph} = _context do
      query = [
        "GRAPH.QUERY",
        graph.name,
        "CREATE (a:actor {name: 'Hugh Jackman'})-[:act]->(m:movie {title:'Wolverine'}) RETURN a",
        "--compact"
      ]

      {:ok, query_result} = RedisGraph.command(conn, query)
      %{result_set: result_set, header: header, statistics: statistics} = query_result

      correct_header = [:a]

      correct_result_set = [
        [
          %RedisGraph.Node{
            id: 0,
            alias: :a,
            labels: ["actor"],
            properties: %{name: "Hugh Jackman"}
          }
        ]
      ]

      assert is_struct(query_result, QueryResult)

      assert header == correct_header
      assert result_set == correct_result_set

      assert Map.get(statistics, "Nodes created") == "2"
      assert Map.get(statistics, "Properties set") == "2"
      assert Map.get(statistics, "Relationships created") == "1"
    end

    test "test that error is being returned", %{conn: conn, graph: graph} = _context do
      query = [
        "GRAPH.QUERY",
        graph.name,
        "MATCH (a:actor {name: 'Hugh Jackman'})-[:act]->(m:movie {title:'Wolverine'})",
        "--compact"
      ]

      {:error, error} = RedisGraph.command(conn, query)
      assert is_struct(error, Redix.Error)
    end
  end

  describe "RedisGraph.query" do
    test "test that QueryResult is being returned with correct data",
         %{conn: conn, graph: graph} = _context do
      query =
        "CREATE (a:actor {name: 'Hugh Jackman'})-[:act]->(m:movie {title:'Wolverine'}) RETURN a"

      {:ok, query_result} = RedisGraph.query(conn, graph.name, query)
      %{result_set: result_set, header: header, statistics: statistics} = query_result

      correct_header = [:a]

      correct_result_set = [
        [
          %RedisGraph.Node{
            id: 0,
            alias: :a,
            labels: ["actor"],
            properties: %{name: "Hugh Jackman"}
          }
        ]
      ]

      assert is_struct(query_result, QueryResult)

      assert header == correct_header
      assert result_set == correct_result_set

      assert Map.get(statistics, "Nodes created") == "2"
      assert Map.get(statistics, "Properties set") == "2"
      assert Map.get(statistics, "Relationships created") == "1"
    end

    test "test that error is being returned", %{conn: conn, graph: graph} = _context do
      query = "MATCH (a:actor {name: 'Hugh Jackman'})-[:act]->(m:movie {title:'Wolverine'})"
      {:error, error} = RedisGraph.query(conn, graph.name, query)
      assert is_struct(error, Redix.Error)
    end
  end

  describe "RedisGraph.execution_plan" do
    test "test that QueryResult is being returned with correct data",
         %{conn: conn, graph: graph} = _context do
      create_query =
        "CREATE (a:actor {name: 'Hugh Jackman'})-[:act]->(m:movie {title:'Wolverine'}) RETURN a"

      {:ok, _query_result} = RedisGraph.query(conn, graph.name, create_query)

      match_query =
        "MATCH (a:actor {name: 'Hugh Jackman'})-[:act]->(m:movie {title:'Wolverine'}) RETURN a"

      {:ok, query_result} = RedisGraph.execution_plan(conn, graph.name, match_query)

      response = query_result

      correct_response = [
        "Results",
        "    Project",
        "        Filter",
        "            Conditional Traverse | (a)->(m:movie)",
        "                Filter",
        "                    Node By Label Scan | (a:actor)"
      ]

      assert response == correct_response
    end

    test "test that error is being returned", %{conn: conn, graph: graph} = _context do
      create_query =
        "CREATE (a:actor {name: 'Hugh Jackman'})-[:act]->(m:movie {title:'Wolverine'}) RETURN a"

      {:ok, _query_result} = RedisGraph.query(conn, graph.name, create_query)

      match_query = "MATCH (a:actor {name: 'Hugh Jackman'})-[:act]->(m:movie {title:'Wolverine'})"
      {:error, error} = RedisGraph.execution_plan(conn, graph.name, match_query)
      assert is_struct(error, Redix.Error)
    end
  end

  describe "RedisGraph.delete" do
    test "test that QueryResult is being returned with correct data",
         %{conn: conn, graph: graph} = _context do
      create_query =
        "CREATE (a:actor {name: 'Hugh Jackman'})-[:act]->(m:movie {title:'Wolverine'}) RETURN a"

      {:ok, _query_result} = RedisGraph.query(conn, graph.name, create_query)
      {:ok, delete_result} = RedisGraph.delete(conn, graph.name)

      assert is_binary(
               Map.get(delete_result.statistics, "Graph removed, internal execution time")
             )
    end

    test "test that error is being returned", %{conn: conn, graph: graph} = _context do
      create_query =
        "CREATE (a:actor {name: 'Hugh Jackman'})-[:act]->(m:movie {title:'Wolverine'}) RETURN a"

      {:ok, _query_result} = RedisGraph.query(conn, graph.name, create_query)
      {:error, error} = RedisGraph.delete(conn, "1234")
      assert is_struct(error, Redix.Error)
    end
  end

  describe "RedisGraph.call_procedure" do
    test "test that QueryResult is being returned with correct data",
         %{conn: conn, graph: graph} = _context do
      query =
        "CREATE (a:actor {name: 'Hugh Jackman'})-[:act]->(m:movie {title:'Wolverine'}) RETURN a"

      {:ok, _query_result} = RedisGraph.query(conn, graph.name, query)

      {:ok, query_result} = RedisGraph.call_procedure(conn, graph.name, "db.labels")
      %{result_set: result_set, header: header} = query_result

      correct_header = [:label]
      correct_result_set = [["actor"], ["movie"]]
      assert is_struct(query_result, QueryResult)
      assert header == correct_header
      assert result_set == correct_result_set
    end

    test "test that error is being returned", %{conn: conn, graph: graph} = _context do
      query =
        "CREATE (a:actor {name: 'Hugh Jackman'})-[:act]->(m:movie {title:'Wolverine'}) RETURN a"

      {:ok, _query_result} = RedisGraph.query(conn, graph.name, query)
      {:error, error} = RedisGraph.call_procedure(conn, graph.name, "test")
      assert is_struct(error, Redix.Error)
    end
  end

  describe "RedisGraph.call_procedure_raw" do
    test "test that QueryResult is being returned with correct data",
         %{conn: conn, graph: graph} = _context do
      query =
        "CREATE (a:actor {name: 'Hugh Jackman'})-[:act]->(m:movie {title:'Wolverine'}) RETURN a"

      {:ok, _query_result} = RedisGraph.query(conn, graph.name, query)
      {:ok, response} = RedisGraph.call_procedure_raw(conn, graph.name, "db.labels")
      assert is_list(response)
    end

    test "test that error is being returned", %{conn: conn, graph: graph} = _context do
      query =
        "CREATE (a:actor {name: 'Hugh Jackman'})-[:act]->(m:movie {title:'Wolverine'}) RETURN a"

      {:ok, _query_result} = RedisGraph.query(conn, graph.name, query)
      {:error, error} = RedisGraph.call_procedure_raw(conn, graph.name, "test")
      assert is_struct(error, Redix.Error)
    end
  end
end


mix.exs
defmodule RedisGraph.MixProject do
  use Mix.Project

  @description "A RedisGraph client library in Elixir with support for Cypher query building."
  @repo_url "https://github.com/AlexSandro19/redisgraph-ex-lib"
  @website_url "https://hexdocs.pm/ex_redisgraph/RedisGraph.html"
  @version "0.1.0"

  def project do
    [
      app: :ex_redisgraph,
      version: @version,
      elixir: "~> 1.14",
      start_permanent: Mix.env() == :prod,
      test_coverage: [tool: ExCoveralls],
      deps: deps(),
      # Docs
      description: @description,
      package: package(),
      source_url: @repo_url,
      homepage_url: @website_url,
      docs: docs()
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  defp package do
    [
      maintainers: ["Alexandru Sandrovschii"],
      licenses: ["MIT"],
      links: %{"GitHub" => @repo_url}
    ]
  end

  defp docs do
    [
      main: "RedisGraph",
      source_ref: "v#{@version}",
      source_url: @repo_url,
      extras: [
        "README.md",
        "CHANGELOG.md"
      ]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:redix, "~> 1.2"},
      {:castore, ">= 0.0.0"},
      {:ex_doc, "~> 0.27", only: :dev, runtime: false},
      {:dialyxir, "~> 1.3", only: [:dev, :test], runtime: false},
      {:excoveralls, "~> 0.16.1", only: [:dev, :test], runtime: false},
      {:credo, "~> 1.7", only: [:dev, :test], runtime: false}
    ]
  end
end


lib/redis_graph.ex
defmodule RedisGraph do
  @moduledoc """

  Query builder library that provides functions
  to construct Cypher queries to communicate with [RedisGraph](https://redis.io/docs/stack/graph/) database
  and interact with the entities through defined structures.

  The library is developed on top of an existing [library](https://github.com/crflynn/redisgraph-ex),
  written by Christopher Flynn, and provides additional functionality with refactored codebase to support
  [RedisGraph result set](https://redis.io/docs/stack/graph/design/client_spec/).

  To run `RedisGraph` locally with Docker, use

  ```bash
  docker run -p 6379:6379 -it --rm redis/redis-stack-server
  ```

  Here is a simple example of how to use the library:

  ```elixir
  alias RedisGraph.{Query, Graph, QueryResult}

  # Create a connection using Redix
  {:ok, conn} = Redix.start_link("redis://localhost:6379")

  # Create a graph
  graph = Graph.new(%{
    name: "social"
  })

  {:ok, query} =
        Query.new()
        |> Query.create()
        |> Query.node(:n, ["Person"], %{age: 30, name: "John Doe", works: true})
        |> Query.relationship_from_to(:r, "TRAVELS_TO", %{purpose: "pleasure"})
        |> Query.node(:m, ["Place"], %{name: "Japan"})
        |> Query.return(:n)
        |> Query.return_property(:n, "age", :Age)
        |> Query.return(:m)
        |> Query.build_query()

  # query will hold
  # "CREATE (n:Person {age: 30, name: 'John Doe', works: true})-[r:TRAVELS_TO {purpose: 'pleasure'}]->(m:Place {name: 'Japan'}) RETURN n, n.age AS Age, m

  # Execute the query
  {:ok, query_result} = RedisGraph.query(conn, graph.name, query)

  # Get result set
  result_set = Map.get(query_result, :result_set)
  # result_set will hold
  # [
  #   [
  #     %RedisGraph.Node{
  #       id: 2,
  #       alias: :n,
  #       labels: ["Person"],
  #       properties: %{age: 30, name: "John Doe", works: true}
  #     },
  #     30,
  #     %RedisGraph.Node{
  #       id: 3,
  #       alias: :m,
  #       labels: ["Place"],
  #       properties: %{name: "Japan"}
  #     }
  #   ]
  # ]

  ```
  """

  alias RedisGraph.{QueryResult, Util}

  require Logger

  @type connection() :: GenServer.server()

  @doc """
  Execute arbitrary command against the database.

  https://redis.io/docs/stack/graph/commands/

  Query commands will be a list of strings. They
  will begin with either `GRAPH.QUERY`,
  `GRAPH.EXPLAIN`, `GRAPH.DELETE` etc.

  The next element will be the name of the graph.

  The third element will be the query command.

  Optionally pass the last element `--compact`
  for compact results.

  Returns a `RedisGraph.QueryResult` containing the result set
  and metadata associated with the query or error message.

  ## Example:
  ```
  alias RedisGraph.Graph

  # Create a connection using Redix
  {:ok, conn} = Redix.start_link("redis://localhost:6379")

  # Create a graph
  graph = Graph.new(%{name: "imdb"})

  # Create the query
  query = [
      "GRAPH.QUERY",
      graph.name,
      "MATCH (a:actor)-[:act]->(m:movie {title:'straight outta compton'}) RETURN a",
      "--compact"
    ]

  # Call the command
  {:ok, query_result} = RedisGraph.command(conn, query)
  ```
  """
  @spec command(connection(), list(String.t())) ::
          {:ok, QueryResult.t()} | {:error, any()}
  def command(conn, c) do
    # Logger.debug(Enum.join(c, " "))

    case Redix.command(conn, c) do
      {:ok, result} ->
        {:ok, QueryResult.new(%{conn: conn, graph_name: Enum.at(c, 1), raw_result_set: result})}

      {:error, _reason} = error ->
        error
    end
  end

  @doc """
  Run a query on a graph in the database.

  Returns a `RedisGraph.QueryResult` containing the result set
  and metadata associated with the query or error message.

  https://redis.io/commands/graph.query/

  ## Example:
  ```
  alias RedisGraph.Graph

  # Create a connection using Redix
  {:ok, conn} = Redix.start_link("redis://localhost:6379")

  # Create a graph
  graph = Graph.new(%{name: "imdb"})

  # Create the query
  query = "MATCH (a:actor)-[:act]->(m:movie {title:'straight outta compton'}) RETURN a"

  # Call the query
  {:ok, query_result} = RedisGraph.query(conn, graph.name, query)
  ```
  """
  @spec query(connection(), String.t(), String.t()) ::
          {:ok, QueryResult.t()} | {:error, any()}
  def query(conn, graph_name, q) do
    c = ["GRAPH.QUERY", graph_name, q, "--compact"]
    command(conn, c)
  end

  @doc """
  Fetch the execution plan for a query on a graph.

  Returns a raw result containing the query plan.

  https://redis.io/commands/graph.explain/

  ## Example:
  ```
  alias RedisGraph.Graph

  # Create a connection using Redix
  {:ok, conn} = Redix.start_link("redis://localhost:6379")

  # Create a graph
  graph = Graph.new(%{name: "imdb"})

  # Create the query
  query = "MATCH (a:actor)-[:act]->(m:movie {title:'straight outta compton'}) RETURN a"

  # Call the query
  {:ok, query_result} = RedisGraph.execution_plan(conn, graph.name, query)
  ```
  """
  @spec execution_plan(connection(), String.t(), String.t()) :: {:ok, list()} | {:error, any()}
  def execution_plan(conn, graph_name, q) do
    c = ["GRAPH.EXPLAIN", graph_name, q]

    case Redix.command(conn, c) do
      {:error, _reason} = error ->
        error

      {:ok, result} ->
        # Logger.debug(result)
        {:ok, result}
    end
  end

  @doc """
  Delete a graph from the database.

  Returns a `RedisGraph.QueryResult` containing the result set
  and metadata associated with the query or error message.

  https://redis.io/commands/graph.delete/

  ## Example:
  ```
  alias RedisGraph.Graph

  # Create a connection using Redix
  {:ok, conn} = Redix.start_link("redis://localhost:6379")

  # Create a graph
  graph = Graph.new(%{name: "imdb"})

  # Call the query
  {:ok, query_result} = RedisGraph.delete(conn, graph.name)
  ```
  """
  @spec delete(connection(), String.t()) ::
          {:ok, QueryResult.t()} | {:error, any()}
  def delete(conn, graph_name) do
    command = ["GRAPH.DELETE", graph_name]
    RedisGraph.command(conn, command)
  end

  @doc """
  Execute a procedure call against the graph specified and receive the raw result of the procedure call.

  https://redis.io/docs/stack/graph/design/client_spec/#procedure-calls

  ## Example:
  ```
  alias RedisGraph.Graph

  # Create a connection using Redix
  {:ok, conn} = Redix.start_link("redis://localhost:6379")

  # Create a graph
  graph = Graph.new(%{name: "imdb"})

  # Call the query
  {:ok, query_result} = RedisGraph.call_procedure_raw(conn, graph.name, "db.labels")
  ```
  """
  @spec call_procedure_raw(connection(), String.t(), String.t(), list(), map()) ::
          {:ok, list()} | {:error, any()}
  def call_procedure_raw(conn, graph_name, procedure, args \\ [], kwargs \\ %{}) do
    args = Enum.map_join(args, ",", &Util.value_to_string/1)

    yields = Map.get(kwargs, "y", [])

    yields =
      if length(yields) > 0 do
        " YIELD " <> Enum.join(yields, ",")
      else
        ""
      end

    q = "CALL " <> procedure <> "(" <> args <> ")" <> yields
    c = ["GRAPH.QUERY", graph_name, q, "--compact"]

    case Redix.command(conn, c) do
      {:ok, result} ->
        {:ok, result}

      {:error, _reason} = error ->
        error
    end
  end

  @doc """
  Execute a procedure call against the graph specified.

  Returns a `RedisGraph.QueryResult` containing the result set
  and metadata associated with the query or error message.

  https://redis.io/docs/stack/graph/design/client_spec/#procedure-calls

  ## Example:
  ```
  alias RedisGraph.Graph

  # Create a connection using Redix
  {:ok, conn} = Redix.start_link("redis://localhost:6379")

  # Create a graph
  graph = Graph.new(%{name: "imdb"})

  # Call the query
  {:ok, query_result} = RedisGraph.call_procedure(conn, graph.name, "db.labels")
  ```
  """
  @spec call_procedure(connection(), String.t(), String.t(), list(), map()) ::
          {:ok, QueryResult.t()} | {:error, any()}
  def call_procedure(conn, graph_name, procedure, args \\ [], kwargs \\ %{}) do
    args = Enum.map_join(args, ",", &Util.value_to_string/1)

    yields = Map.get(kwargs, "y", [])

    yields =
      if length(yields) > 0 do
        " YIELD " <> Enum.join(yields, ",")
      else
        ""
      end

    q = "CALL " <> procedure <> "(" <> args <> ")" <> yields
    c = ["GRAPH.QUERY", graph_name, q, "--compact"]

    RedisGraph.command(conn, c)
  end

  @doc """
  Fetch response of the `db.labels()` procedure call against the specified graph.
  It returns a `RedisGraph.QueryResult` containing the result set and metadata from the call.
    ## Example:
  ```
  alias RedisGraph.Graph

  # Create a connection using Redix
  {:ok, conn} = Redix.start_link("redis://localhost:6379")

  # Create a graph
  graph = Graph.new(%{name: "imdb"})

  # Call the query
  {:ok, query_result} = RedisGraph.labels(conn, graph.name)
  ```
  """
  @spec labels(connection(), String.t()) :: {:ok, list()} | {:error, any()}
  def labels(conn, graph_name) do
    call_procedure(conn, graph_name, "db.labels")
  end

  @doc """
  Fetch response of the `db.relationshipTypes()` procedure call against the specified graph.
  It returns a `RedisGraph.QueryResult` containing the result set and metadata from the call.
    ## Example:
  ```
  alias RedisGraph.Graph

  # Create a connection using Redix
  {:ok, conn} = Redix.start_link("redis://localhost:6379")

  # Create a graph
  graph = Graph.new(%{name: "imdb"})

  # Call the query
  {:ok, query_result} = RedisGraph.relationship_types(conn, graph.name)
  ```
  """
  @spec relationship_types(connection(), String.t()) :: {:ok, list()} | {:error, any()}
  def relationship_types(conn, graph_name) do
    call_procedure(conn, graph_name, "db.relationshipTypes")
  end

  @doc """
  Fetch response of the `db.propertyKeys()` procedure call against the specified graph.
  It returns a `RedisGraph.QueryResult` containing the result set and metadata from the call.
    ## Example:
  ```
  alias RedisGraph.Graph

  # Create a connection using Redix
  {:ok, conn} = Redix.start_link("redis://localhost:6379")

  # Create a graph
  graph = Graph.new(%{name: "imdb"})

  # Call the query
  {:ok, query_result} = RedisGraph.property_keys(conn, graph.name)
  ```
  """
  @spec property_keys(connection(), String.t()) :: {:ok, list()} | {:error, any()}
  def property_keys(conn, graph_name) do
    call_procedure(conn, graph_name, "db.propertyKeys")
  end
end
